operators/nftopology/controllers/nftopology_controller.go

```diff
@@ -0,0 +1,182 @@
+type NFTopologyReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	l      logr.Logger
+}
+
+func AllUPFsDeployed(neighborMap map[string][]reqv1alpha1.NFInstance, nfInst *reqv1alpha1.NFInstance, nfDeployed *deployv1alpha1.NFDeployed, deployedInstMap map[string]int) bool {
+
+	expected := 0
+	count := 0
+	if neighborList, ok := neighborMap[nfInst.Name]; !ok {
+		return false
+	} else {
+		// create a map for fast lookup
+		neighborLookupMap := map[string]struct{}{}
+		for _, neighbor := range neighborList {
+			if neighbor.NFTemplate.NFType == reqv1alpha1.NFTypeUPF {
+				neighborLookupMap[neighbor.Name] = struct{}{}
+				expected++
+			}
+		}
+
+		for neighborInstName, idx := range deployedInstMap {
+			if _, ok := neighborLookupMap[GetNFInstanceNameFromNFDeployedName(neighborInstName)]; ok {
+				if nfDeployed.Spec.NFInstances[idx].NFType == string(reqv1alpha1.NFTypeUPF) {
+					count++
+				}
+			}
+		}
+	}
+
+	if expected == count {
+		return true
+	} else {
+		return false
+	}
+}
+
+/*
+ * BuildPVS: builds a PVS object from NFTopology's NFInstance
+ * NFClass needed for package reference
+ */
+func BuildPVS(nfInst *reqv1alpha1.NFInstance, nfClassObj *reqv1alpha1.NFClass) *pvsv1alpha2.PackageVariantSet {
+	retPVS := &pvsv1alpha2.PackageVariantSet{}
+
+	retPVS.ObjectMeta.Name = nfInst.Name
+	upstream := &pvv1alpha1.Upstream{}
+	upstream.Repo = nfClassObj.Spec.PackageRef.RepositoryName
+	upstream.Package = nfClassObj.Spec.PackageRef.PackageName
+	upstream.Revision = nfClassObj.Spec.PackageRef.Revision
+
+	retPVS.Spec.Upstream = upstream
+
+	target := &pvsv1alpha2.Target{}
+	objectSelector := &pvsv1alpha2.ObjectSelector{}
+
+	labelMap := map[string]string{}
+	for k, v := range nfInst.ClusterSelector.MatchLabels {
+		labelMap[k] = v
+	}
+
+	objectSelector.LabelSelector = metav1.LabelSelector{MatchLabels: labelMap}
+	objectSelector.APIVersion = WORKLOADAPIVERSION
+	objectSelector.Kind = WORKLOADKIND
+	target.ObjectSelector = objectSelector
+
+	retPVS.Spec.Targets = append(retPVS.Spec.Targets, *target)
+
+	// TODO(s3wong): adding Template in PVS?
+	return retPVS
+}
+
+//+kubebuilder:rbac:groups=req.nephio.org,resources=nftopologies,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=req.nephio.org,resources=nftopologies/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=req.nephio.org,resources=nftopologies/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *NFTopologyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.l = log.FromContext(ctx).WithValues("req", req)
+	r.l.Info("Reconcile NFTopology")
+
+	nfTopo := &reqv1alpha1.NFTopology{}
+	if err := r.Client.Get(ctx, req.NamespacedName, nfTopo); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	nfDeployed := &deployv1alpha1.NFDeployed{}
```

Comment:
I'm still a little confused as to what NFDeployed is. It almost seems like it should be part of NFTopology.Status. Or is the idea to have many of them for one NFTopology, in which case I can see using a separate resource.